src/QsCompiler/Tests.Compiler/TestCases/QirTests/TestOpArgument.ll

```diff
@@ -22,6 +24,8 @@ entry:
   call void @__quantum__qis__cnot__body(%Qubit* %q1, %Qubit* %q2)
   call void @__quantum__qis__swap(%Qubit* %q1, %Qubit* %q2)
   %8 = call %Callable* @__quantum__rt__callable_create([4 x void (%Tuple*, %Tuple*, %Tuple*)*]* @Microsoft__Quantum__Intrinsic__CNOT__FunctionTable, [2 x void (%Tuple*, i32)*]* null, %Tuple* null)
```

Comment:
This doesn't appear to be a new problem, but some of these callabes look like they are getting leaked. The callable `%8` is created (ref count: 1), new code inserts an increment because it gets stored into a tuple (ref count: 2), then new code tracks when the tuple references go away (ref count: 1). Nothing takes away that final ref count. In the old QIR from this diff, the ref count on `%8` was never managed at all, so the the leak seems to be preexisting.

Compare this to `%40` below which gets created (ref count: 1), passed as an argument, and then decremented (ref count: 0), so that it is not leaked.

Reply:
There is line 33 
```llvm
store %Callable* %8, %Callable** %11, align 8
```
that stores the callable, so I believe a ref count should be correct. We do validate that there are no memory leaks as part of the e2e (Q# samples), and while I don't know if this exact case is covered, it would be rather surprising if it wasn't.

Reply:
I did get the chance to run this case through the evaluator and saw that no memory leaks are detected, so you are right, I just can't see how. I'll get into detail below, but again, it looks like the pattern than concerns me exists before this change, so I don't think it's any reason to worry about this change (hence the sign off). Here's my logic for future reference though...

Looking at the file _before_ the changes, I can see that a callable is created on line 24 that is stored in value `%8`:
https://github.com/microsoft/qsharp-compiler/blob/7714168fda4379fb7e6a6c616f680ec039c482f4/src/QsCompiler/Tests.Compiler/TestCases/QirTests/TestOpArgument.ll#L24
At this point, it has reference count of 1 by definition of being newly created. It is then immediately stored in the tuple created as value `%9` and accessed through `%11` (tuple also has ref count of 1):
https://github.com/microsoft/qsharp-compiler/blob/7714168fda4379fb7e6a6c616f680ec039c482f4/src/QsCompiler/Tests.Compiler/TestCases/QirTests/TestOpArgument.ll#L25-L28
After that, `%8` and `%11` are never used again, so any access would have to come through the tuple `%9` or it's bitcast value `%10`. However, `%10` is also never used again outside of these lines, and `%9` is only used one more time as an argument to the creation of another callable value on line 31:
https://github.com/microsoft/qsharp-compiler/blob/7714168fda4379fb7e6a6c616f680ec039c482f4/src/QsCompiler/Tests.Compiler/TestCases/QirTests/TestOpArgument.ll#L31-L32
This callable value `%13` is used as an argument to a function call right below that, and then later cleaned up by decrementing the alias and reference count at the end of the function:
https://github.com/microsoft/qsharp-compiler/blob/7714168fda4379fb7e6a6c616f680ec039c482f4/src/QsCompiler/Tests.Compiler/TestCases/QirTests/TestOpArgument.ll#L81-L82
But the original callable value in `%8` and tuple value in `%9` are _never_ interacted with again, so I can't see how/where the original reference count would be decremented to zero and the two items cleaned up.

The only difference in this specific flow after the changes made here is an extra increment and decrement. So instead of appearing to have a value of 1 throughout execution, `%8` is immediately incremented to ref count of 2 after creation:
https://github.com/microsoft/qsharp-compiler/blob/6745f31d11b1fa8d2ceb400524256bc6ea35800c/src/QsCompiler/Tests.Compiler/TestCases/QirTests/TestOpArgument.ll#L26-L28
and then decremented back to 1 at the end of the function:
https://github.com/microsoft/qsharp-compiler/blob/6745f31d11b1fa8d2ceb400524256bc6ea35800c/src/QsCompiler/Tests.Compiler/TestCases/QirTests/TestOpArgument.ll#L95-L96
But there still doesn't appear to be anything that would decrement the final ref count and clean up the value. Likewise, the access pattern for the tuple in `%9` hasn't changed at all; it is created, stored into, passed as an argument to another callable creation, and then never accessed again, so appears to reach the end of the function with a ref count of 1.

It's very likely I'm missing something here that makes this all work, given that taking this exact code and putting into an execution tests passes the leak checks, but the pattern in the IR itself seems suspicious.